api/versions.go

```diff
@@ -369,56 +369,62 @@ func (api *DatasetAPI) updateVersion(ctx context.Context, body io.ReadCloser, ve
-		if err != nil {
-			log.Error(ctx, "putVersion endpoint: failed to model version resource based on request", err, data)
-			return nil, nil, nil, errs.ErrUnableToParseJSON
-		}
+	versionUpdate, err := models.CreateVersion(body, versionDetails.datasetID)
+	if err != nil {
+		log.Error(ctx, "putVersion endpoint: failed to model version resource based on request", err, data)
+		return nil, nil, nil, errs.ErrUnableToParseJSON
+	}
 
-		currentDataset, err := api.dataStore.Backend.GetDataset(ctx, versionDetails.datasetID)
-		if err != nil {
-			log.Error(ctx, "putVersion endpoint: datastore.getDataset returned an error", err, data)
-			return nil, nil, nil, err
-		}
+	currentDataset, err := api.dataStore.Backend.GetDataset(ctx, versionDetails.datasetID)
+	if err != nil {
+		log.Error(ctx, "putVersion endpoint: datastore.getDataset returned an error", err, data)
+		return nil, nil, nil, err
+	}
 
-		if err = api.dataStore.Backend.CheckEditionExists(ctx, versionDetails.datasetID, versionDetails.edition, ""); err != nil {
-			log.Error(ctx, "putVersion endpoint: failed to find edition of dataset", err, data)
-			return nil, nil, nil, err
-		}
+	if err = api.dataStore.Backend.CheckEditionExists(ctx, versionDetails.datasetID, versionDetails.edition, ""); err != nil {
+		log.Error(ctx, "putVersion endpoint: failed to find edition of dataset", err, data)
+		return nil, nil, nil, err
+	}
 
-		currentVersion, err := api.dataStore.Backend.GetVersion(ctx, versionDetails.datasetID, versionDetails.edition, version, "")
-		if err != nil {
-			log.Error(ctx, "putVersion endpoint: datastore.GetVersion returned an error", err, data)
-			return nil, nil, nil, err
-		}
+	currentVersion, err := api.dataStore.Backend.GetVersion(ctx, versionDetails.datasetID, versionDetails.edition, version, "")
+	if err != nil {
+		log.Error(ctx, "putVersion endpoint: datastore.GetVersion returned an error", err, data)
+		return nil, nil, nil, err
+	}
 
-		// Combine update version document to existing version document
-		populateNewVersionDoc(currentVersion, versionUpdate)
-		data["updated_version"] = versionUpdate
-		log.Info(ctx, "putVersion endpoint: combined current version document with update request", data)
+	// acquire instance lock to make sure we read the correct values of dimension options
+	lockID, err := api.dataStore.Backend.AcquireInstanceLock(ctx, currentVersion.ID)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	defer api.dataStore.Backend.UnlockInstance(ctx, lockID)
 
-		if err = models.ValidateVersion(versionUpdate); err != nil {
-			log.Error(ctx, "putVersion endpoint: failed validation check for version update", err)
-			return nil, nil, nil, err
-		}
+	currentVersion, err = api.dataStore.Backend.GetVersion(ctx, versionDetails.datasetID, versionDetails.edition, version, "")
+	if err != nil {
+		log.Error(ctx, "putVersion endpoint: datastore.GetVersion returned an error", err, data)
+		return nil, nil, nil, err
+	}
```

Comment:
This might look like duplicated code to developers coming here in the future. I assume there is no better option than making 2 calls to `GetVersion` but I'd suggest adding some comments detailing why it needs to be done this way (because we don't have the instance id to lock if we don't call GetVersion?)

Reply:
I agree with Rafa, maybe a comment before line 410 that says something like:
// By the time the lock is acquired, the 'currentVersion' may have changed, so we have to get it again

(thats if what i've just suggested is correct)

Reply:
In fact, there might be a way to do a single get call:
- add eTag to Version model (it exists in the instance documents already, which is the collection we are using for versions)
- (acquired)
- read eTag from first call
- update query with eTag selector (if it changed, the eTag will be different, so the query will fail)
- if the query fails, do another get+update.
- (release lock)

This would prevent the 2 calls unless there is a race condition.